### ISSUE TITLE
Completion-based I/O

### ISSUE THREAD

**Issue by Ralith** [quinn-rs/quinn#915]
Modern high-performance I/O APIs (e.g. io_uring and I/O completion ports) are completion-oriented, unlike the traditional readiness-oriented `epoll` paradigm. Advantages include fewer syscalls and no copying. On Windows in particular, readiness-oriented I/O is poorly supported.

@Matthias247 has reported that a prototype variant of Quinn modified to use registered I/O on windows performs significantly (~20%?) better than our Linux backend using `sendmmsg` and `recvmmsg` for efficient batching, and drastically better than the fallback backend on Windows. Due to major changes in tokio/mio's windows support, we should re-evaluate the latter result after moving to tokio 0.3.

On Linux, io_uring is only available on recent kernels. Other platforms (e.g. macOS, BSDs) may not offer completion-based I/O at all. Retaining a readiness-oriented fallback will therefore remain necessary for some time. However, the performance benefits seem to justify making completion-based I/O the first-class target.

A complicating factor is that tokio itself is, currently, 100% readiness-oriented. On Linux, we may be able to bridge this gap gracefully due to io_uring/epoll interop, but it's unclear if something similar is possible on Windows. If not, a background thread may be necessary.

**Comment by Matthias247** [comment]
I pushed my POC for RIO and completion based IO to https://github.com/quinn-rs/quinn/pull/918

This might help on getting an idea what is necessary. I think in a model where Quinn owns all IO and the associated thread it's not terribly complicated.
The fact that datagrams are received and transmitted in an all-or-nothing fashion, and that it mostly has to deal with a single socket and not lots of them makes it easier than e.g. implementing support for completion oriented TCP. But obviously one still has to be a bit careful around ownership.

**Comment by KirillLykov** [comment]
I wonder if there is any plans about using `io_uring`? It looks like another way to proceed might be using AF_XDP which might be simpler(?) in a way

**Comment by Ralith** [comment]
See also discussion at https://github.com/quinn-rs/quinn/issues/1319. This issue was originally opened before GSO/GRO support were implemented and it's unclear if io_uring will be a benefit in net on Linux, though we'd be happy to mentor someone who wants to experiment.

**Comment by KirillLykov** [comment]
> See also discussion at #1319. This issue was originally opened before GSO/GRO support were implemented and it's unclear if io_uring will be a benefit in net on Linux, though we'd be happy to mentor someone who wants to experiment.

It sounds super interesting to check out, not sure if I will manage yet. I would start with  creating a hack to plugin monoio and check the performance metrics. Hence, the first question is if there are any existing benchmarks that I can use to measure the performance. And the second -- where to start from?

**Comment by Ralith** [comment]
See the `bench` and `perf` directories for some benchmarking tools. As discussed in #1319, tokio-uring might be a more appropriate place to start. You could look at @Matthias247's prototype above for inspiration, or try modifying the endpoint driver in `quinn` directly, or try building directly on top of quinn-proto to avoid being influenced by the existing architecture.

**Comment by sowhu** [comment]
I just discovered this from Stephen's talk last Friday. I am also quite interested in io-uring, xdp and network performance in general. What is the current status on this, XDP and GRO/GSO? Just wonder how I can contribute. 
Regarding GSO/GRO, I think it only helps for TCP traffic. I thought we are using UDP. But I fairly new here, so it could be totally wrong.

**Comment by sowhu** [comment]
> @sowhu do you have more context on the talk? Not sure which Stephen you're referring to.

Stephen Akridge gave a talk to the lab folks last week outlining a list of outstanding issues he thinks might be worth visiting. This is one of the them. He also mentioned that we might need to run some experiments to see how much performance gain we can achieve using io_uring.

**Comment by sowhu** [comment]
Sorry guys. I just realized I posted on the wrong issue. Please just disregard my previous two comments. My mistake.

**Comment by Icelk** [comment]
> Those benchmarks don't seem to involve tokio-uring, just regular epoll-based tokio.

I did some benchmarks with tokio-uring (which is single-threaded). It scores 370k/s, while a single-threaded monoio scores 270k/s. Monoio can scale to multiple threads, but so can tokio-uring (if you just spawn multiple executors!).

TL;DR: `tokio-uring` seems to be the way to go.

I would really like to use QUIC with io_uring! Thanks.

**Comment by thomaseizinger** [comment]
I haven't looked at the code or performance at all but I just stumbled over https://github.com/compio-rs/compio/blob/master/compio-quic which uses `quinn-proto` on top of its own completion-based IO runtime.
